File: DSA/DSA_450/Python/14_DP/054_Word_Break.py
"""
Problem: Word Break
URL: https://practice.geeksforgeeks.org/problems/word-break1352/1

Problem Statement:
Given a string s and a dictionary of words, determine if s can be segmented into a space-separated sequence of one or more dictionary words.

Sample Input/Output:
Input: s = "leetcode", dict = ["leet","code"]
Output: true
"""


class TrieNode:
    def __init__(self):
        self.is_end = False
        self.children = [None] * 26


class Solution:
    def Word_Break_DP(self, s: str, word_dict: list[str]) -> bool:
        """
        DP approach
        Time Complexity: O(n^2*L)
        Space Complexity: O(n)
        """
        n = len(s)
        dict_set = set(word_dict)
        dp = [False] * (n + 1)
        dp[0] = True
        
        for i in range(1, n + 1):
            for j in range(i):
                if dp[j] and s[j:i] in dict_set:
                    dp[i] = True
                    break
        
        return dp[n]
    
    def _insert(self, root: TrieNode, word: str) -> None:
        node = root
        for c in word:
            idx = ord(c) - ord('a')
            if not node.children[idx]:
                node.children[idx] = TrieNode()
            node = node.children[idx]
        node.is_end = True
    
    def Word_Break_Trie(self, s: str, word_dict: list[str]) -> bool:
        """
        Trie-based approach
        Time Complexity: O(n^2)
        Space Complexity: O(n + m*L)
        """
        root = TrieNode()
        for word in word_dict:
            self._insert(root, word)
        
        n = len(s)
        dp = [False] * (n + 1)
        dp[0] = True
        
        for i in range(n):
            if not dp[i]:
                continue
            
            node = root
            for j in range(i, n):
                idx = ord(s[j]) - ord('a')
                if not node.children[idx]:
                    break
                
                node = node.children[idx]
                if node.is_end:
                    dp[j + 1] = True
        
        return dp[n]


def Test_WordBreak():
    solution = Solution()
    
    s = "leetcode"
    word_dict = ["leet", "code"]
    
    assert solution.Word_Break_DP(s, word_dict) == True
    assert solution.Word_Break_Trie(s, word_dict) == True


if __name__ == "__main__":
    Test_WordBreak()

File: DSA/DSA_450/Cpp/14_DP/054_Word_Break.cpp
/*
Problem: Word Break
URL: https://practice.geeksforgeeks.org/problems/word-break1352/1

Problem Statement:
Given a string s and a dictionary of words, determine if s can be segmented into a space-separated sequence of one or more dictionary words.

Sample Input/Output:
Input: s = "leetcode", dict = ["leet","code"]
Output: true
*/

#include <bits/stdc++.h>
using namespace std;

class Solution {
public:
    bool Word_Break_DP(string s, vector<string>& wordDict) {
        /*
        DP approach
        Time Complexity: O(n^2*L)
        Space Complexity: O(n)
        */
        int n = s.length();
        unordered_set<string> dict(wordDict.begin(), wordDict.end());
        vector<bool> dp(n + 1, false);
        dp[0] = true;
        
        for (int i = 1; i <= n; i++) {
            for (int j = 0; j < i; j++) {
                if (dp[j] && dict.find(s.substr(j, i - j)) != dict.end()) {
                    dp[i] = true;
                    break;
                }
            }
        }
        
        return dp[n];
    }
    
    struct TrieNode {
        bool isEnd;
        TrieNode* children[26];
        TrieNode() {
            isEnd = false;
            for (int i = 0; i < 26; i++) children[i] = nullptr;
        }
    };
    
    void Insert(TrieNode* root, string& word) {
        TrieNode* node = root;
        for (char c : word) {
            int idx = c - 'a';
            if (!node->children[idx]) {
                node->children[idx] = new TrieNode();
            }
            node = node->children[idx];
        }
        node->isEnd = true;
    }
    
    bool Word_Break_Trie(string s, vector<string>& wordDict) {
        /*
        Trie-based approach
        Time Complexity: O(n^2)
        Space Complexity: O(n + m*L)
        */
        TrieNode* root = new TrieNode();
        for (string& word : wordDict) {
            Insert(root, word);
        }
        
        int n = s.length();
        vector<bool> dp(n + 1, false);
        dp[0] = true;
        
        for (int i = 0; i < n; i++) {
            if (!dp[i]) continue;
            
            TrieNode* node = root;
            for (int j = i; j < n; j++) {
                int idx = s[j] - 'a';
                if (!node->children[idx]) break;
                
                node = node->children[idx];
                if (node->isEnd) {
                    dp[j + 1] = true;
                }
            }
        }
        
        return dp[n];
    }
};

void Test_Word_Break() {
    Solution solution;
    
    string s = "leetcode";
    vector<string> dict = {"leet", "code"};
    
    cout << "DP: s=\"" << s << "\" -> " 
         << (solution.Word_Break_DP(s, dict) ? "true" : "false") << endl;
    cout << "Trie: s=\"" << s << "\" -> " 
         << (solution.Word_Break_Trie(s, dict) ? "true" : "false") << endl;
}

int main() {
    Test_Word_Break();
    return 0;
}
